src/storage/src/hummock/store/version.rs

```diff
@@ -198,30 +207,213 @@ impl StagingVersion {
+                    {
+                        let new_indices =
+                            self.construct_index(sub_level, old_object_ids.iter().copied());
+                        new_levels.push((*old_sub_level_id, new_indices));
+
+                        sub_level_asc_iter.next();
+                    }
+                }
+                None => break,
+            }
+        }
+
+        // Add new sub levels in the latest `HummockVersion` into `self`.
+        let mut levels_to_extend = vec![];
+        for (idx, sub_level) in committed_l0_sub_levels.iter().enumerate() {
+            while let Some((epoch, _)) = cleaned_staging_ssts.last() && epoch > &sub_level.sub_level_id {
+                cleaned_staging_ssts.pop();
+            }
+            if cleaned_staging_ssts.is_empty() {
+                break;
+            }
+
+            let next_epoch = committed_l0_sub_levels
+                .get(idx + 1)
+                .map(|sub_level| sub_level.get_sub_level_id());
+            let left_len = if let Some(next_epoch) = next_epoch {
+                let mut left_len = cleaned_staging_ssts.len();
+                while left_len > 0 && cleaned_staging_ssts[left_len - 1].0 > next_epoch {
+                    left_len -= 1;
+                }
+                left_len
+            } else {
+                0
+            };
+            let candidate_ssts = cleaned_staging_ssts.split_off(left_len);
+
+            if sub_level.level_type() != LevelType::Nonoverlapping {
+                let committed_indices = self.construct_index(
+                    sub_level,
+                    candidate_ssts
+                        .into_iter()
+                        .flat_map(|(_, object_ids)| object_ids),
+                );
+                levels_to_extend.push((sub_level.get_sub_level_id(), committed_indices));
+            }
+        }
+
+        new_levels.extend(levels_to_extend.into_iter().rev());
+
+        // Build result.
+        let mut ret = vec![];
+        for sub_level in committed_l0_sub_levels {
+            let sub_level_id = sub_level.get_sub_level_id();
+            if sub_level_id < self.scope {
+                break;
+            }
+            ret.push(if let Some((new_sub_level_id, _)) = new_levels.last() && *new_sub_level_id == sub_level_id {
+                new_levels.pop().unwrap()
+            } else {
+                (sub_level_id, (vec![], vec![]))
+            });
+        }
+
+        Self {
+            table_id: self.table_id,
+            scope: self.scope,
+            levels: ret,
+        }
+    }
+}
+
 /// A container of information required for reading from hummock.
 pub struct HummockReadVersion {
+    table_id: TableId,
+
+    is_singleton: bool,
+    vnodes: Arc<Bitmap>,
+
     /// Local version for staging data.
     staging: StagingVersion,
 
     /// Remote version for committed data.
     committed: CommittedVersion,
```

Comment:
Both `CommittedVersion` and `PrunedVersion` are a kind of view to the committed data. Is it possible that we unify them and only store a unified version view here? And we can also only use this unified version view in the read version tuple instead of (CommittedVersion + PrunedVersion) so that the read logic can be simplified.

Personally I think this unified version view may look like
```
struct {
     sub_levels: Vec<PrunedLevel>,
     levels: Vec<PrunedLevel>,
     max_committed_epoch: HummockEpoch,
}
```

Reply:
> Both `CommittedVersion` and `PrunedVersion` are a kind of view to the committed data. Is it possible that we unify them and only store a unified version view here? And we can also only use this unified version view in the read version tuple instead of (CommittedVersion + PrunedVersion) so that the read logic can be simplified.
> 
> Personally I think this unified version view may look like
> 
> ```
> struct {
>      sub_levels: Vec<PrunedLevel>,
>      levels: Vec<PrunedLevel>,
>      max_committed_epoch: HummockEpoch,
> }
> ```

`CommittedVersion` is a shared `Arc` so if we unify them, we will do many unnecessary clones.